Add manage pipeline components content to user guide

ptitzler opened the pull request:
Closes #1873
~Requires #2083~
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our [contributor guidelines](https://github.com/elyra-ai/community/blob/master/contributing.md)
    1.1 Please follow the [7 rules for a great commit message](https://github.com/elyra-ai/community/blob/master/contributing.md#creating-a-pull-request)
  2. If the PR is unfinished, leave it as 'DRAFT', add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...' or use the 'status:Work in Progress' label.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If this PR involves UI changes, please provide a screen capture (before/after) for a faster review.
  6. Remember to add 'Fixes #ISSUE_NUMBER' (or any [valid keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)) to the end of your message body to get the issue properly closed when the PR is merged.
  7. Set the proper milestone based on the target release for the issue.
  8. Consider whether any documentation need to be added or updated based on your changes.
-->

### What changes were proposed in this pull request?
- Update _Pipeline components_ topic in the user guide https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html mimicking  https://elyra.readthedocs.io/en/stable/user_guide/runtime-image-conf.html
  ![image](https://user-images.githubusercontent.com/13068832/134315712-2277c0dc-128e-4bc8-aa03-bc4716837f16.png)
- Update _Command line interface_ topic in the user guide
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor code that leads to class changes, showing the updated class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
-->

### How was this pull request tested?

Reviewed output of `make docs`
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly
including negative and positive cases if possible.

If it was tested in a way different from regular unit tests, please clarify how you tested step by step,
ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.

If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ptitzler commented:
I won't be able to deliver this in time for 3.1.0.

ptitzler commented:
@kiersten-stokes is it possible for a user to create `filename` and `directory` registry entries without having to rebuild elyra?

kiersten-stokes commented:
@ptitzler Currently there are two options for `filename` and `directory`-based registries:

1. The user enters an absolute path (this is what the hints/description suggests in the metadata editor right now). This can include `~` to represent the home directory (via `os.path.expanduser`)
2. The user can add components to the `etc/config/components/[runtime]` folder(s). This will populate them in the "Preloaded" registry for that runtime and **does** require a rebuild of elyra.

We haven't added support for any other relative base directory yet, but we could pretty easily add some logic to the existing location reading method where first an absolute path is checked for, then relative, then finally checking the ENV_JUPYTER_PATH for the components added on build. I'm guessing we would want to use something like the `notebook-dir` as the base directory for relative paths? But I suppose there would have to be an understanding that if you change your `notebook-dir` that those components then won't show up anymore

ptitzler commented:
@kiersten-stokes are `etc/config/components/kfp` and `etc/config/components/airflow` by any chance currently hardwired in the code?  I thought that we just happened to name the directories after the runtimes, but that does not seem to be the case based on the following test:
 - cloned elyra source
 - created a directory `elyra/etc/config/components/abc`
 - added a KFP YAML file to that directory
 - built elyra
 - added a `Directory` registry entry for KFP by setting `Paths` to `abc`. 
    ```
    $ elyra-metadata list component-registries
     ...
     component-registry   hello                             /Users/patti/Library/Jupyter/metadata/component-registries/hello.json 

     $ cat /Users/patti/Library/Jupyter/metadata/component-registries/hello.json
      {
       "display_name": "hello",
       "metadata": {
       "runtime": "kfp",
       "location_type": "Directory",
       "paths": [
         "abc"
       ]
     },
     "schema_name": "component-registry"
    }

    $ pwd
     /Users/patti/elyra_dev/upstream/elyra/etc/config/components

    $ ls -R
    abc	airflow	kfp

    ./abc:
     something.yaml

    ./airflow:
    slack_operator.py

    ./kfp:
    filter_text_using_shell_and_grep.yaml	something.yaml
    run_notebook_using_papermill.yaml	something_else.yaml
    ```

 - restarted elyra
 
The console log contains the following warning message and the component does not appear in the palette:
``` 
[W 2021-09-20 14:39:17.746 ServerApp] Invalid directory -> /opt/anaconda3/envs/upstream/share/jupyter/components/abc```

kiersten-stokes commented:
> @kiersten-stokes are `etc/config/components/kfp` and `etc/config/components/airflow` by any chance currently hardwired in the code? I thought that we just happened to name the directories after the runtimes, but that does not seem to be the case based on the following test:

They are currently hardcoded, yes. `setup.py` copies:
`components_kfp_path = './etc/config/components/kfp/*.yaml'` into `'share/jupyter/components/kfp'` and
`components_airflow_path = './etc/config/components/airflow/*.py'` into `'share/jupyter/components/airflow'`. 

I haven't tested it out, but don't see an obvious reason why we wouldn't be able to support additional directories in `etc/config` if we thought there would be value in that

kevin-bates commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -50,18 +50,186 @@ Component details and demo pipelines can be found in the `https://github.com/ely
+
+##### N/A (name)
+
+A unique identifier for this registry entry. A value is generated from `Name` if no value is provided.
```
"A unique identifier" sounds like a UUID.  I think something more like "The canonical name for this registry entry" might be slightly more correct.

ptitzler replied:
changed!

kevin-bates commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -50,18 +50,186 @@ Component details and demo pipelines can be found in the `https://github.com/ely
+
+##### Runtime (runtime)
+
+The runtime environment that supports the component(s). Valid values are `kfp` and `airflow`. This property is required.
```
This doesn't account for other runtimes or uninstalled runtimes.  I think we should state something like: _Valid values are the set of currently configured runtimes that appear in the dropdown (UI) or help-text (CLI)._

ptitzler replied:
Changed! That said (off-topic), I do believe the CLI should also provide a command that lists supported runtimes (user friendly name and  value to be specified). (Help is still hard to use)

Example (ignoring for now semantic ambiguity because elyra-metadata uses `runtimes` to refer to runtime configurations:

```
$ elyra-pipeline list runtimes
   runtime id       runtime name
   ---------------------------------
   kfp                   Kubeflow Pipelines
   airflow             Apache Airflow 1.x
```

kevin-bates commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -50,18 +50,186 @@ Component details and demo pipelines can be found in the `https://github.com/ely
+- `URL`: The provided `Paths` identify web resources. The pipeline editor loads the specified URLs using anonymous HTTP `GET` requests.
+- `Filename`: The provided `Paths` identify files in the file system where JupyterLab/Elyra is running. `~` may be used to denote the user's home directory.
+
+The `Directory` location type is used to _add components during the Elyra build process_. The provided `Paths` must identify existing directories in [`etc/config/components`](https://github.com/elyra-ai/elyra/tree/master/etc/config/components).
```
I don't think the _build-time_ aspect of this is correct.  `Directory` should be a location where users can dump their component specifications and independent of the "build process".  That directory can reside anywhere on the system (provided users have appropriate access rights).  If I'm understanding things, we should keep `etc/config/components` to ourselves (i.e., it is reserved for the _build-time_ needs we have to deliver _factory assets_) and the only "non-absolute" paths allowed are the two from the factory: `airflow` and `kfp`.  (Again, if I'm understanding things correctly.)

`Directory` should then be another bullet in the list and _just an option_ like `URL` and `Filename` - there's nothing special about it other than it implies all files in its location are component definitions.

ptitzler replied:
Changed and removed references to relative path [for now].

kevin-bates commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -50,18 +50,186 @@ Component details and demo pipelines can be found in the `https://github.com/ely
+ - URL: `['https://raw.githubusercontent.com/elyra-ai/elyra/master/etc/config/components/kfp/run_notebook_using_papermill.yaml']`
+ - Filename: `['/Users/patti/specs/load_data_from_public_source/http_operator.py']`
+ - Filename: `['~patti/specs/filter_files/row_filter.yaml']`
+ - Directory path: `['load_from_database']`
```
Any user-provided registry should always contain absolute path references.

ptitzler replied:
fixed!

kevin-bates commented on docs/source/user_guide/pipeline-components.md:
```diff
@@ -50,18 +50,186 @@ Component details and demo pipelines can be found in the `https://github.com/ely
+
+ Examples multiple components (CLI):
+ - URL: `['URL1', 'URL2']`
+ - Filename: `['/Users/patti/specs/comp1.yaml','/Users/patti/specs/comp2.yaml']`
```
Seems like a multi-valued example for `Directory` is warranted...
```
- Directory: `['/Users/patti/specs','/Users/patti/specs2']`
```

ptitzler replied:
added!

kevin-bates reviewed (commented):
Just had some wording comments and, in general, it seems like we need to firm up what exactly the "Directory" location_type means.

ptitzler commented:
[off topic] Noticed during testing that more server messages `[I...]` are logged

```
(upstream) ~/elyra_dev/doc-manage-components/elyra $ elyra-metadata install component-registries \
       --name="filter_components" \
       --display_name="filter components" \
       --description="filter text in files" \
       --runtime=kfp \
       --location_type=Directory \
       --paths="['/Users/patti/elyra_dev/doc-manage-components/workspace/kfp-component-tests/component_definitions/count-lines', '~/elyra_dev/doc-manage-components/workspace/kfp-component-tests/component_definitions/download-file']" \
       --categories='["file operations"]' \
       --replace
[I 2021-09-22 11:06:13,598.598] Registering airflow processor "elyra.pipeline.airflow.processor_airflow.AirflowPipelineProcessor"...
[I 2021-09-22 11:06:13,633.633] Registering kfp processor "elyra.pipeline.kfp.processor_kfp.KfpPipelineProcessor"...
[I 2021-09-22 11:06:13,714.714] Registering local processor "elyra.pipeline.local.processor_local.LocalPipelineProcessor"...
Metadata instance 'filter_components' for schema 'component-registry' has been written to: /Users/patti/Library/Jupyter/metadata/component-registries/filter_components.json
```

ptitzler commented:
Addressed all review comments.

kevin-bates commented:
> [off topic] Noticed during testing that more server messages [I...] are logged

These are logged when the entrypoints are loaded as a result of constructing the pipeline processor registry.  They come from the server, not the application.  About all we could do is convert them to DEBUG statements, although I personally find them useful to see when the Elyra server extension is starting in Lab and feel INFO is the correct log level, but also understand their "noise" within a CLI application.

Hmm, there might be a way to detect the application at the point of the logging statement and determine if the execution thread is there on behalf of Lab or a CLI, but it might also be cleaner to convert these to DEBUG as the former can be a slippery slope.

kevin-bates reviewed (approved):
LGTM - thanks Patrick.  We can address the logging topic separately.
